components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state.cc

```diff
@@ -0,0 +1,123 @@
+
+P3ABandwidthSavingsPermanentState::P3ABandwidthSavingsPermanentState(
+    PrefService* user_prefs)
+    : user_prefs_(user_prefs) {
+  if (user_prefs) {
+    LoadSavings();
+  }
+}
+
+P3ABandwidthSavingsPermanentState::~P3ABandwidthSavingsPermanentState() =
+    default;
+
+void P3ABandwidthSavingsPermanentState::AddSavings(uint64_t delta) {
+  base::Time now_midnight = base::Time::Now().LocalMidnight();
+  base::Time last_saved_midnight;
+
+  if (!daily_savings_.empty()) {
+    last_saved_midnight = daily_savings_.front().day;
+  }
+
+  if (now_midnight - last_saved_midnight > base::TimeDelta()) {
+    // Day changed.
+    daily_savings_.push_front({now_midnight, delta});
+    if (daily_savings_.size() > kNumOfSavedDailyUptimes) {
+      daily_savings_.pop_back();
+    }
+  } else {
+    daily_savings_.front().saving += delta;
+  }
+
+  RecordP3A();
+  SaveSavings();
+}
+
+uint64_t P3ABandwidthSavingsPermanentState::GetTotalSavings() const {
+  // We record only saving for last N days.
+  const base::Time n_days_ago =
+      base::Time::Now() - base::TimeDelta::FromDays(kNumOfSavedDailyUptimes);
+  return std::accumulate(daily_savings_.begin(), daily_savings_.end(),
+                         DailySaving(),
+                         [n_days_ago](const auto& u1, const auto& u2) {
+                           uint64_t add = 0;
+                           // Check only last continious days.
+                           if (u2.day > n_days_ago) {
+                             add = u2.saving;
+                           }
+                           return DailySaving{{}, u1.saving + add};
+                         })
+      .saving;
+}
+
+void P3ABandwidthSavingsPermanentState::LoadSavings() {
+  DCHECK(daily_savings_.empty());
+  const base::ListValue* list =
+      user_prefs_->GetList(prefs::kBandwidthSavedDailyBytes);
+  if (!list) {
+    return;
+  }
+  for (auto it = list->begin(); it != list->end(); ++it) {
+    const base::Value* day = it->FindKey("day");
+    const base::Value* saving = it->FindKey("saving");
+    if (!day || !saving || !day->is_double() || !saving->is_double()) {
+      continue;
+    }
+    if (daily_savings_.size() == kNumOfSavedDailyUptimes) {
+      break;
+    }
+    daily_savings_.push_back({base::Time::FromDoubleT(day->GetDouble()),
+                              (uint64_t)saving->GetDouble()});
+  }
+}
+
+void P3ABandwidthSavingsPermanentState::SaveSavings() {
+  DCHECK(!daily_savings_.empty());
+  DCHECK_LE(daily_savings_.size(), kNumOfSavedDailyUptimes);
+
+  ListPrefUpdate update(user_prefs_, prefs::kBandwidthSavedDailyBytes);
+  base::ListValue* list = update.Get();
+  list->Clear();
+  for (const auto& u : daily_savings_) {
+    base::DictionaryValue value;
+    value.SetKey("day", base::Value(u.day.ToDoubleT()));
+    value.SetKey("saving", base::Value(static_cast<double>(u.saving)));
+    list->GetList().push_back(std::move(value));
+  }
+}
+
+void P3ABandwidthSavingsPermanentState::RecordP3A() {
```

Comment:
this is used for p3a, but it's just writing a histogram so the convention would be `RecordSavingsDaily`

Reply:
it also helps to avoid confusion about whether this needs a p3a guard or not which is why I was looking more closely at it